Sources/SwiftProtobuf/BinaryDecoder.swift

```diff
@@ -1007,11 +1007,22 @@ internal struct BinaryDecoder: Decoder {
 
     internal mutating func decodeExtensionField(values: inout ExtensionFieldValueSet, messageType: Message.Type, fieldNumber: Int) throws {
         if let ext = extensions?[messageType, fieldNumber] {
-            var fieldValue = values[fieldNumber] ?? ext._protobuf_newField()
             assert(!consumed)
-            try fieldValue.decodeExtensionField(decoder: &self)
+            var fieldValue = values[fieldNumber]
+            if fieldValue != nil {
+                try fieldValue!.decodeExtensionField(decoder: &self)
+            } else {
+                fieldValue = try ext._protobuf_newField(decoder: &self)
+            }
             if consumed {
-              values[fieldNumber] = fieldValue
+                if fieldValue != nil {
+                    values[fieldNumber] = fieldValue
+                } else {
+                    // Really things should never get here as the required data
+                    // should always be available or the decode of the extension
+                    // would throw.
+                    throw BinaryDecodingError.truncated
```

Comment:
I debated that, but the only part haven't fully unwound is what would happen on a binary blob if someone crafted what should be a repeated-packed-pod extension, and then on the wire sent the tag and a length of zero for the data.  I don't think that would blow up in the wire decode, and I currently pick off no values being decoded in `init?`, but that can't directly throw because it doesn't know which type of decoding error to throw.  Things like that have always sorta been a grey area in the spec, it shouldn't happen, but I also can't detect it at the field decode level because the empty list case can happen if the wire type was wrong and that concept is specific to the binary decoder.

I don't think any of the optional extensions could ever end up here as they should always throw their format specific error when they can't extract data (or consumed here would be false for wrong wire types).

We could change the repeated decoded to never return nil, as when doing a merging decode (the other decode method in the ExtensionFields), there is no check to ensure they did decode something.

Should I just tweak the two comments to capture some of these ramblings instead?